mailpoet/lib/Automation/Integrations/MailPoet/Fields/SubscriberCustomFieldsFactory.php

```diff
@@ -0,0 +1,166 @@
+      default:
+        throw new InvalidStateException(sprintf('Unknown custom field type "%s"', $customField->getType()));
+    }
+  }
+
+  private function createStringField(CustomFieldEntity $customField): Field {
+    $factory = function (SubscriberPayload $payload) use ($customField) {
+      return $this->getCustomFieldValue($payload, $customField);
+    };
+    return $this->createField($customField, Field::TYPE_STRING, $factory);
+  }
+
+  private function createNumberField(CustomFieldEntity $customField): Field {
+    $factory = function (SubscriberPayload $payload) use ($customField) {
+      $value = $this->getCustomFieldValue($payload, $customField);
+      return is_numeric($value) ? (float)$value : null;
+    };
+    return $this->createField($customField, Field::TYPE_NUMBER, $factory);
+  }
+
+  private function createBooleanField(CustomFieldEntity $customField): Field {
+    $factory = function (SubscriberPayload $payload) use ($customField) {
+      $value = $this->getCustomFieldValue($payload, $customField);
+      return $value === null ? null : (bool)$value;
+    };
+    return $this->createField($customField, Field::TYPE_BOOLEAN, $factory);
+  }
+
+  private function createEnumField(CustomFieldEntity $customField): Field {
+    $factory = function (SubscriberPayload $payload) use ($customField) {
+      $value = $this->getCustomFieldValue($payload, $customField);
+      return $value === null ? null : $value;
+    };
+    return $this->createField($customField, Field::TYPE_ENUM, $factory, [
+      'options' => array_map(function (array $value) {
+        return ['id' => $value['value'], 'name' => $value['value']];
+      }, $customField->getParams()['values'] ?? []),
+    ]);
+  }
+
+  private function createDateTimeField(CustomFieldEntity $customField): Field {
+    $factory = function (SubscriberPayload $payload) use ($customField) {
+      return $this->getDateTimeValue($customField, $this->getCustomFieldValue($payload, $customField) ?? '');
+    };
+    return $this->createField($customField, Field::TYPE_DATETIME, $factory);
+  }
+
+  private function createYearField(CustomFieldEntity $customField): Field {
+    $factory = function (SubscriberPayload $payload) use ($customField) {
+      $value = $this->getDateTimeValue($customField, $this->getCustomFieldValue($payload, $customField) ?? '');
+      return $value ? (int)$value->format('Y') : null;
+    };
+    return $this->createField($customField, Field::TYPE_INTEGER, $factory);
+  }
+
+  private function createMonthField(CustomFieldEntity $customField): Field {
+    $factory = function (SubscriberPayload $payload) use ($customField) {
+      $value = $this->getDateTimeValue($customField, $this->getCustomFieldValue($payload, $customField) ?? '');
+      return $value ? (int)$value->format('n') : null;
+    };
+    return $this->createField($customField, Field::TYPE_ENUM, $factory, [
+      'options' => array_map(function (int $value) {
+        global $wp_locale;
```

Comment:
Can't we inject `\WP_Locale`? Or is `change_locale` a problem here?

Reply:
@websupporter Hmm, not sure. Might be safer to use the "freshest" possible value?
What do you think about adding this to our `WordPress` wrapper service:

```php
public function getWpLocale(): WP_Locale {
  global $wp_locale;
  return $wp_locale;
}
```

Reply:
I like that